fgpyo/vcf/builder.py

```diff
@@ -0,0 +1,378 @@
+        if sd is None:
+            sd = VariantBuilder.default_sd()
+        # add mandatory VCF format
+        yield "##fileformat=VCFv4.2"
+        # add GT
+        yield '##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">'
+        # add additional common INFO lines
+        yield '##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">'
+        yield (
+            '##INFO=<ID=AR,Number=A,Type=Float,Description="Allele Ratio - ratio of AD for allele'
+            ' vs. AD for modal allele.">'
+        )
+        yield '##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">'
+        # add additional common FORMAT lines
+        yield (
+            '##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt'
+            ' alleles in the order listed">'
+        )
+        yield '##FORMAT=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">'
+        yield '##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Total Depth">'
+
+        for d in sd.values():
+            if "ID" not in d or "length" not in d:
+                raise ValueError(
+                    "Sequence dictionary must include 'ID' and 'length' for each contig."
+                )
+            contig_id = d["ID"]
+            contig_length = d["length"]
+            contig_header = f"##contig=<ID={contig_id},length={contig_length}"
+            for key, value in d.items():
+                if key == "ID" or key == "length":
+                    continue
+                contig_header += f",{key}={value}"
+            contig_header += ">"
+            yield contig_header
+
+    @property
+    def num_samples(self) -> int:
+        return len(self.sample_ids)
+
+    def add(
+        self,
+        contig: Optional[str] = None,
+        pos: int = 1000,
+        id: str = ".",
+        ref: str = "A",
+        alts: Union[None, str, Iterable[str]] = (".",),
+        qual: int = 60,
+        filter: Optional[Iterable[str]] = None,
+        info: Optional[Dict[str, Any]] = None,
+        samples: Optional[Dict[str, Dict[str, Any]]] = None,
+    ) -> VariantRecord:
+        """Generates a new variant and adds it to the internal collection.
+
+        Notes:
+        * Very little validation is done with respect to INFO and FORMAT keys being defined in the
+        header.
+        * VCFs are 1-based, but pysam is (mostly) 0-based. We define the function in terms of the
+        VCF property "pos", which is 1-based. pysam will also report "pos" as 1-based, so that is
+        the property that should be accessed when using the records produced by this function (not
+        "start").
+
+        Args:
+            contig: the chromosome name. If None, will use the first contig in the sequence
+                    dictionary.
+            pos: the 1-based position of the variant
+            id: the variant id
+            ref: the reference allele
+            alts: the list of alternate alleles, None if no alternates
```

Comment:
Given the type could you add a little more.  If `str` is that considered to be a singular ALT?